torchdata/datapipes/iter/util/cacheholder.py

```diff
@@ -177,11 +177,11 @@ def _hash_check(self, filepath):
 
         return hash_func.hexdigest() == self.hash_dict[filepath]  # type: ignore[index]
 
-    def _end_caching(self):
+    def _end_caching(self) -> IterDataPipe:
         todo_dp, cached_dp = self.source_datapipe.demux(2, self._cache_check_fn)
         self._end_caching_flag = True
         # Re-assign source_datapipe
-        self.source_datapipe = todo_dp
+        # self.source_datapipe = todo_dp  # TODO: This causes infinite recursion in graph traversal
```

Comment:
@ejguan If we do not reassign to `source_datapipe`, there will not be infinite recursion when the DataPipe graph is traversed.

One potential issue that I can think of may be that `source_datapipe` may advance without being in sync with `cached_dp`. If that is an issue, we should have a test to catch that. What do you think?